components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java

```diff
@@ -119,13 +122,60 @@ private String[] getAppAssociatedRolesForLocalUser(AuthenticatedUser authenticat
         Set<String> userRoleIds = getAllRolesOfLocalUser(authenticatedUser);
         List<RoleV2> rolesAssociatedWithApp = getRolesAssociatedWithApplication(applicationId,
                 authenticatedUser.getTenantDomain());
+        if (StringUtils.isNotEmpty(authenticatedUser.getSharedUserId())) {
+            // Add the shared role details to the roles list which are associated with the application.
+            addSharedRoleAssociations(authenticatedUser, rolesAssociatedWithApp, userRoleIds);
+        }
 
         return rolesAssociatedWithApp.stream()
                 .filter(role -> userRoleIds.contains(role.getId()))
                 .map(role -> appendInternalDomain(role.getName()))
                 .toArray(String[]::new);
     }
 
+    private void addSharedRoleAssociations(AuthenticatedUser authenticatedUser, List<RoleV2> rolesAssociatedWithApp,
+                                           Set<String> userRoleIds) throws ApplicationRolesException {
+
+        if (!isSharedUserAccessingSharedOrg(authenticatedUser)) {
+            return;
+        }
```

Comment:
This logic looks redundant as we are invoking this method for organization switched users.